File: Yixuan/18_01_17/divide_two_integers.py
class Solution:
    # @param {int} dividend the dividend
    # @param {int} divisor the divisor
    # @return {int} the result
    # @link: http://www.lintcode.com/en/problem/divide-two-integers/
    # @author: Egbert Li
    def divide(self, dividend, divisor):
        INT_MAX = 2147483647
        a = dividend > 0 ? dividend : -dividend
        b = divisor > 0 ? divisor : -divisor
        result, shift = 0, 31
        while shift >= 0:
            if a >= (b << shift):
                a -= b << shift
                result += b << shift
            shift -= 1
        if (dividend ^ divisor) >> 31:
            result = -result
        if result > INT_MAX:
            return INT_MAX
        return result
        

File: Yixuan/18_01_17/divide_two_integers.cpp
class Solution {
public:
    /**
     * @param dividend the dividend
     * @param divisor the divisor
     * @return the result
     * @author: Egbert Li
     * @link: http://www.lintcode.com/en/problem/divide-two-integers/
     */
   int divide(int dividend, int divisor) {
      long long a = dividend > 0 ? dividend : - (long long) dividend;
      long long b = divisor > 0 ? divisor : - (long long) divisor;
      long long result = 0, shift = 31;
      while (a >= b) {
         if (a >= (b << shift)) {
            a -= b << (shift);
            result += 1LL << (shift);
         }
         shift --;
      }
      result = (dividend ^ divisor) >> 31 ? -result : result;
      if (result > INT32_MAX) {
         return INT32_MAX;
      }
      return result;
   }
};
